server/src/main/java/io/crate/execution/engine/aggregation/impl/average/AverageState.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Licensed to Crate.io GmbH ("Crate") under one or more contributor
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.execution.engine.aggregation.impl.average;
+
+import io.crate.types.ByteType;
+import io.crate.types.DataType;
+import io.crate.types.DoubleType;
+import io.crate.types.FloatType;
+import io.crate.types.IntegerType;
+import io.crate.types.LongType;
+import io.crate.types.ShortType;
+import io.crate.types.TimestampType;
+
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Objects;
+
+public abstract class AverageState implements Comparable<AverageState>, Writeable {
+
+    protected double sum = 0;
+    protected long count = 0L;
+
+    public static AverageState of(DataType<?> dataType) {
+        switch (dataType.id()) {
+            case DoubleType.ID:
+            case FloatType.ID:
+                return new NonIntegralAverageState();
+            case LongType.ID:
+            case IntegerType.ID:
+            case ShortType.ID:
+            case ByteType.ID:
+            case TimestampType.ID_WITH_TZ:
+                return new IntegralAverageState();
+            default:
+                throw new IllegalArgumentException(String.format(Locale.ENGLISH, "data type %s is not supported", dataType.getName()));
+        }
+    }
```

Comment:
There is one place where I used it and I moved switch there - if I got comment correctly(remove type handing from AverageState). I didn't create a supplier since it would be same amount of code and it's used only in `public AverageState newState`

Reply:
My thought was that the registration is already concrete anyways

```
            for (var supportedType : SUPPORTED_TYPES) {
                mod.register(
                    Signature.aggregate(
                        functionName,
                        supportedType.getTypeSignature(),
                        DataTypes.DOUBLE.getTypeSignature()),
                    AverageAggregation::new
                );
            }
```

We could split this into fractional/integral and already provide the right type factory instead of asking about the type instance after the fact.

Reply:
got it, did one more commit